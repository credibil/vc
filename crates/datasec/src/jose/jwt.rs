//! # JSON Web Token (JWT)
//!
//! JSON Web Token (JWT) is a compact, URL-safe means of representing
//! claims to be transferred between two parties.  The claims in a JWT
//! are encoded as a JSON object that is used as the payload of a JSON
//! Web Signature (JWS) structure or as the plaintext of a JSON Web
//! Encryption (JWE) structure, enabling the claims to be digitally
//! signed or integrity protected with a Message Authentication Code
//! (MAC) and/or encrypted.

use std::fmt::{Debug, Display};

use serde::{Deserialize, Serialize};

use crate::jose::jwk::PublicKeyJwk;
use crate::Algorithm;

/// Represents a JWT as used for proof and credential presentation.
#[derive(Clone, Debug, Default, Serialize, PartialEq, Eq)]
pub struct Jwt<T> {
    /// The JWT header.
    pub header: Header,

    /// The JWT claims.
    pub claims: T,
}

/// Represents the JWT header.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Header {
    /// Digital signature algorithm identifier as per IANA "JSON Web Signature
    /// and Encryption Algorithms" registry.
    pub alg: Algorithm,

    /// Used to declare the media type [IANA.MediaTypes](http://www.iana.org/assignments/media-types)
    /// of the JWS.
    pub typ: Type,

    /// The key material for the public key.
    #[serde(flatten)]
    pub key: KeyType,

    /// Contains a certificate (or certificate chain) corresponding to the key
    /// used to sign the JWT. This element MAY be used to convey a key
    /// attestation. In such a case, the actual key certificate will contain
    /// attributes related to the key properties.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub x5c: Option<String>,

    /// Contains an OpenID.Federation Trust Chain. This element MAY be used to
    /// convey key attestation, metadata, metadata policies, federation
    /// Trust Marks and any other information related to a specific
    /// federation, if available in the chain.
    ///
    /// When used for signature verification, `kid` MUST be set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trust_chain: Option<String>,
}

/// The JWT `typ` claim.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub enum Type {
    /// JWT `typ` for Verifiable Credential.
    #[default]
    #[serde(rename = "jwt")]
    Credential,

    /// JWT `typ` for Verifiable Presentation.
    #[serde(rename = "jwt")]
    Presentation,

    /// JWT `typ` for Authorization Request Object.
    #[serde(rename = "oauth-authz-req+jwt")]
    Request,

    /// JWT `typ` for Wallet's Proof of possession of key material.
    #[serde(rename = "openid4vci-proof+jwt")]
    Proof,
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

/// The type of public key material for the JWT.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub enum KeyType {
    /// Contains the key ID. If the Credential is bound to a DID, the kid refers
    /// to a DID URL which identifies a particular key in the DID Document
    /// that the Credential should bound to. Alternatively, may refer to a
    /// key inside a JWKS.
    #[serde(rename = "kid")]
    KeyId(String),

    /// Contains the key material the new Credential shall be bound to.
    #[serde(rename = "jwk")]
    Jwk(PublicKeyJwk),
}

impl Default for KeyType {
    fn default() -> Self {
        Self::KeyId(String::new())
    }
}
