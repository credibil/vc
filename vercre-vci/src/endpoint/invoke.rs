//! # Credential Offer Handler
//!
//! The Credential Offer endpoint is for Credential Issuers to generate a Credential Offer
//! which can be used to initiate Credential issuance with a Wallet.
//!
//! When a Credential Issuer is already interacting with a user and wishes to initate a
//! Credential issuance, they can 'send' the user's Wallet a Credential Offer.
//!
//! The diagram illustrates this Credential Issuer initiated flow:
//!
//! ```text
//! +--------------+   +-----------+                                    +-------------------+
//! | User         |   |   Wallet  |                                    | Credential Issuer |
//! +--------------+   +-----------+                                    +-------------------+
//!         |                |                                                    |
//!         |                |  (1) User provides  information required           |
//!         |                |      for the issuance of a certain Credential      |
//!         |-------------------------------------------------------------------->|
//!         |                |                                                    |
//!         |                |  (2) Credential Offer (Pre-Authorized Code)        |
//!         |                |<---------------------------------------------------|
//!         |                |  (3) Obtains Issuer's Credential Issuer metadata   |
//!         |                |<-------------------------------------------------->|
//!         |   interacts    |                                                    |
//!         |--------------->|                                                    |
//!         |                |                                                    |
//!         |                |  (4) Token Request (Pre-Authorized Code, pin)      |
//!         |                |--------------------------------------------------->|
//!         |                |      Token Response (access_token)                 |
//!         |                |<---------------------------------------------------|
//!         |                |                                                    |
//!         |                |  (5) Credential Request (access_token, proof(s))   |
//!         |                |--------------------------------------------------->|
//!         |                |      Credential Response                           |
//!         |                |      (credential(s))                               |
//!         |                |<---------------------------------------------------|
//! ```
//!
//! While JSON-based, the Offer can be sent to the Wallet's Credential Offer Handler URL
//! as an HTTP GET request, an HTTP redirect, or a QR code.
//!
//! Below is a non-normative example of a Credential Offer Object for a Pre-Authorized
//! Code Flow (with a credential type reference):
//!
//! ```json
//! {
//!    "credential_issuer": "https://credential-issuer.example.com",
//!    "credentials": [
//!       "UniversityDegree_LDP_VC"
//!    ],
//!    "grants": {
//!       "urn:ietf:params:oauth:grant-type:pre-authorized_code": {
//!           "pre-authorized_code": "adhjhdjajkdkhjhdj",
//!           "user_pin_required": true
//!       }
//!   }
//! }
//! ```
//!
//! See <https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-credential-offer-endpoint>

use std::fmt::Debug;

use anyhow::anyhow;
use chrono::Utc;
use tracing::{instrument, trace};
use vercre_core::error::Err;
use vercre_core::metadata::Issuer as IssuerMetadata;
use vercre_core::vci::{
    AuthorizationCodeGrant, CredentialOffer, Grants, InvokeRequest, InvokeResponse,
    PreAuthorizedCodeGrant,
};
use vercre_core::{
    err, gen, Callback, Client, Holder, Issuer, Result, Server, Signer, StateManager,
};

use super::Handler;
use crate::state::{AuthState, Expire, State};

/// Credential Offer request handler.
impl<P> Handler<P, InvokeRequest>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
{
    /// Call the request for the Request Object endpoint.
    #[instrument]
    pub async fn call(&self) -> Result<InvokeResponse> {
        trace!("Handler::call");
        self.handle_request(Context::new()).await
    }
}

#[derive(Debug)]
struct Context<P>
where
    P: Issuer + StateManager,
{
    callback_id: Option<String>,
    issuer_meta: IssuerMetadata,
    provider: Option<P>,
}

impl<P> Context<P>
where
    P: Issuer + StateManager,
{
    #[instrument]
    pub fn new() -> Self {
        trace!("Context::new");
        Self {
            callback_id: None,
            issuer_meta: IssuerMetadata::default(),
            provider: None,
        }
    }
}

impl<P> vercre_core::Context for Context<P>
where
    P: Issuer + StateManager + Debug,
{
    type Provider = P;
    type Request = InvokeRequest;
    type Response = InvokeResponse;

    // Prepare the context for processing the request.
    #[instrument]
    async fn init(&mut self, req: &Self::Request, provider: Self::Provider) -> Result<&Self> {
        trace!("Context::prepare");

        self.callback_id = req.callback_id.clone();
        self.issuer_meta = Issuer::metadata(&provider, &req.credential_issuer).await?;
        self.provider = Some(provider);

        Ok(self)
    }

    fn callback_id(&self) -> Option<String> {
        self.callback_id.clone()
    }

    // Verify the request is valid.
    #[instrument]
    async fn verify(&self, req: &Self::Request) -> Result<&Self> {
        trace!("Context::verify");

        // credential_issuer required
        if req.credential_issuer.is_empty() {
            err!(Err::InvalidRequest, "No credential_issuer specified");
        };
        // credentials required
        if req.credentials.is_empty() {
            err!(Err::InvalidRequest, "No credentials requested");
        };
        // requested credential is supported
        for cred_id in &req.credentials {
            let Some(_) = self.issuer_meta.credentials_supported.get(cred_id) else {
                err!(Err::UnsupportedCredentialType, "Requested credential is unsupported");
            };
        }
        // holder_id is required
        if req.holder_id.is_none() {
            err!(Err::InvalidRequest, "No holder_id specified");
        };

        Ok(self)
    }

    // Process the request.
    #[instrument]
    async fn process(&self, req: &Self::Request) -> Result<Self::Response> {
        trace!("Context::process");

        let Some(provider) = &self.provider else {
            err!("Provider not set");
        };

        let mut state = State::builder()
            .credential_issuer(req.credential_issuer.clone())
            .expires_at(Utc::now() + Expire::AuthCode.duration())
            .credentials(req.credentials.clone())
            .holder_id(req.holder_id.clone())
            .callback_id(req.callback_id.clone())
            .build();

        let mut pre_auth_grant = None;
        let mut auth_grant = None;
        let mut user_pin = None;

        if req.pre_authorize {
            // ------------------------------------------------
            // Pre-authorized Code Grant
            // ------------------------------------------------
            let pre_auth_code = gen::auth_code();

            pre_auth_grant = Some(PreAuthorizedCodeGrant {
                pre_authorized_code: pre_auth_code.clone(),
                user_pin_required: Some(req.user_pin_required),
                interval: None,
                authorization_server: None,
            });

            if req.user_pin_required {
                user_pin = Some(gen::user_pin());
            }

            // save state by pre-auth_code
            state.auth = Some(
                AuthState::builder()
                    // .issuer_state(iss_state.clone())
                    .user_pin(user_pin.clone())
                    .build(),
            );
            StateManager::put(provider, &pre_auth_code, state.to_vec(), state.expires_at).await?;
        } else {
            // ------------------------------------------------
            // Authorization Code Grant
            // ------------------------------------------------
            let iss_state = gen::state_key();

            auth_grant = Some(AuthorizationCodeGrant {
                issuer_state: Some(iss_state.clone()),
                authorization_server: None,
            });
            StateManager::put(provider, &iss_state, state.to_vec(), state.expires_at).await?;
        }

        // return response
        Ok(InvokeResponse {
            credential_offer: Some(CredentialOffer {
                credential_issuer: req.credential_issuer.clone(),
                credentials: req.credentials.clone(),
                grants: Some(Grants {
                    authorization_code: auth_grant,
                    pre_authorized_code: pre_auth_grant,
                }),
            }),
            user_pin,

            // LATER: save offer to state and return uri
            credential_offer_uri: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use assert_let_bind::assert_let;
    use insta::assert_yaml_snapshot as assert_snapshot;
    use serde_json::json;
    use test_utils::vci_provider::{Provider, ISSUER, NORMAL_USER};

    use super::*;

    #[tokio::test]
    async fn pre_authorize() {
        test_utils::init_tracer();

        let provider = Provider::new();

        // create offer to 'send' to the app
        let body = json!({
            "credentials": ["EmployeeID_JWT"],
            "holder_id": NORMAL_USER,
            "pre-authorize": true,
            "user_pin_required": true,
            "callback_id": "1234"
        });

        let mut request =
            serde_json::from_value::<InvokeRequest>(body).expect("request should deserialize");
        request.credential_issuer = ISSUER.to_string();
        let response = Handler::new(&provider, request).call().await.expect("response is ok");
        assert_snapshot!("invoke", response, {
            ".credential_offer.grants.authorization_code.issuer_state" => "[state]",
            ".credential_offer.grants[\"urn:ietf:params:oauth:grant-type:pre-authorized_code\"][\"pre-authorized_code\"]" => "[pre-authorized_code]",
            ".user_pin" => "[user_pin]"
        });

        // check redacted fields
        let offer = response.credential_offer.as_ref().expect("has credential_offer");
        assert_let!(Some(grants), &offer.grants);
        assert_let!(Some(pre_auth_code), &grants.pre_authorized_code);
        assert!(grants.pre_authorized_code.is_some());

        // compare response with saved state
        let state_key = &pre_auth_code.pre_authorized_code; //as_ref().expect("has state");
        let buf = StateManager::get(&&provider, state_key).await.expect("state exists");
        let state = State::try_from(buf).expect("state is valid");

        assert_snapshot!("state", state, {
            ".expires_at" => "[expires_at]",
            ".auth.code"=>"[code]",
            ".auth.issuer_state" => "[issuer_state]",
            ".auth.user_pin" => "[user_pin]"
        });

        assert_let!(Some(auth_state), &state.auth);
        assert_eq!(auth_state.user_pin, response.user_pin);
    }
}
