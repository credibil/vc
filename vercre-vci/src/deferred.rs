//! # Deferred Credential Endpoint
//!
//! This endpoint is used to issue a Credential previously requested at the Credential
//! Endpoint or Batch Credential Endpoint in cases where the Credential Issuer was not
//! able to immediately issue this Credential.
//!
//! The Wallet MUST present to the Deferred Endpoint an Access Token that is valid for
//! the issuance of the Credential previously requested at the Credential Endpoint or
//! the Batch Credential Endpoint.

use std::fmt::Debug;

use anyhow::anyhow;
use tracing::{instrument, trace};
use vercre_core::error::Err;
#[allow(clippy::module_name_repetitions)]
pub use vercre_core::vci::{DeferredCredentialRequest, DeferredCredentialResponse};
use vercre_core::{err, Callback, Client, Holder, Issuer, Result, Server, Signer, StateManager};

use super::Endpoint;
use crate::state::State;

impl<P> Endpoint<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
{
    /// Deferred credential request handler.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    pub async fn deferred(
        &self, request: &DeferredCredentialRequest,
    ) -> Result<DeferredCredentialResponse> {
        let Ok(buf) = StateManager::get(&self.provider, &request.access_token).await else {
            err!(Err::AccessDenied, "invalid access token");
        };
        let Ok(state) = State::try_from(buf) else {
            err!(Err::AccessDenied, "invalid state for access token");
        };

        let ctx = Context {
            callback_id: state.callback_id.clone(),
        };

        self.handle_request(request, ctx).await
    }
}

#[derive(Debug)]
struct Context {
    callback_id: Option<String>,
}

impl super::Context for Context {
    type Request = DeferredCredentialRequest;
    type Response = DeferredCredentialResponse;

    // TODO: get callback_id from state
    fn callback_id(&self) -> Option<String> {
        self.callback_id.clone()
    }

    #[instrument]
    async fn process<P>(&self, provider: &P, request: &Self::Request) -> Result<Self::Response>
    where
        P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
    {
        trace!("Context::process");

        // retrieve deferred credential request from state
        let Ok(buf) = StateManager::get(provider, &request.transaction_id).await else {
            err!(Err::InvalidTransactionId, "deferred state not found");
        };
        let Ok(state) = State::try_from(buf) else {
            err!(Err::InvalidTransactionId, "deferred state is expired or corrupted");
        };

        let Some(deferred_state) = state.deferred else {
            err!("Deferred state not found.");
        };

        // remove deferred state item
        StateManager::purge(provider, &request.transaction_id).await?;

        // make credential request
        let mut cred_req = deferred_state.credential_request;
        cred_req.credential_issuer = request.credential_issuer.clone();
        cred_req.access_token = request.access_token.clone();

        let response = Endpoint::new(provider.clone()).credential(&cred_req).await?;

        Ok(DeferredCredentialResponse {
            credential_response: response,
        })
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use assert_let_bind::assert_let;
    use base64ct::{Base64UrlUnpadded, Encoding};
    use chrono::Utc;
    use insta::assert_yaml_snapshot as assert_snapshot;
    use serde_json::json;
    use test_utils::vci_provider::{Provider, ISSUER, NORMAL_USER};
    use test_utils::wallet;
    use vercre_core::jwt::{self, Jwt};
    use vercre_core::vci::{CredentialRequest, ProofClaims};
    use vercre_core::w3c::vc::VcClaims;

    use super::*;
    use crate::state::{DeferredState, Expire, TokenState};

    #[tokio::test]
    async fn deferred_ok() {
        test_utils::init_tracer();

        let provider = Provider::new();
        let access_token = "tkn-ABCDEF";
        let c_nonce = "1234ABCD".to_string();
        let transaction_id = "txn-ABCDEF";
        let credentials = vec!["EmployeeID_JWT".to_string()];

        // create CredentialRequest to 'send' to the app
        let jwt_enc = Jwt {
            header: jwt::Header {
                typ: "vercre-vci-proof+jwt".to_string(),
                alg: wallet::alg(),
                kid: wallet::kid(),
            },
            claims: ProofClaims {
                iss: wallet::did(),
                aud: ISSUER.to_string(),
                iat: Utc::now().timestamp(),
                nonce: c_nonce.to_string(),
            },
        }
        .to_string();
        let sig = wallet::sign(jwt_enc.as_bytes());
        let sig_enc = Base64UrlUnpadded::encode_string(&sig);
        let signed_jwt = format!("{jwt_enc}.{sig_enc}");

        let body = json!({
            "credential_identifier": "EmployeeID_JWT",
            "proof":{
                "proof_type": "jwt",
                "jwt": signed_jwt
            }
        });

        let mut cred_req =
            serde_json::from_value::<CredentialRequest>(body).expect("request should deserialize");
        cred_req.credential_issuer = ISSUER.to_string();
        cred_req.access_token = access_token.to_string();

        // set up state
        let mut state = State::builder()
            .credential_issuer(ISSUER.to_string())
            .expires_at(Utc::now() + Expire::AuthCode.duration())
            .credential_configuration_ids(credentials)
            .holder_id(Some(NORMAL_USER.to_string()))
            .build()
            .expect("should build state");

        // state entry 1: token state keyed by access_token
        state.token = Some(TokenState {
            access_token: access_token.to_string(),
            token_type: "Bearer".to_string(),
            c_nonce,
            c_nonce_expires_at: Utc::now() + Expire::Nonce.duration(),
            ..Default::default()
        });
        StateManager::put(&provider, access_token, state.to_vec(), state.expires_at)
            .await
            .expect("state exists");

        // state entry 2: deferred state keyed by transaction_id
        state.token = None;
        state.deferred = Some(DeferredState {
            transaction_id: transaction_id.to_string(),
            credential_request: cred_req.clone(),
        });
        StateManager::put(&provider, transaction_id, state.to_vec(), state.expires_at)
            .await
            .expect("state exists");

        let request = DeferredCredentialRequest {
            credential_issuer: ISSUER.to_string(),
            access_token: access_token.to_string(),
            transaction_id: transaction_id.to_string(),
        };

        let response =
            Endpoint::new(provider.clone()).deferred(&request).await.expect("response is valid");
        assert_snapshot!("response", response, {
            ".credential" => "[credential]",
            ".c_nonce" => "[c_nonce]",
            ".c_nonce_expires_in" => "[c_nonce_expires_in]"
        });

        // extract credential response
        let cred_resp = response.credential_response;

        // verify credential
        let vc_val = cred_resp.credential.expect("VC is present");
        let vc_b64 = serde_json::from_value::<String>(vc_val).expect("base64 encoded string");
        let vc_jwt = Jwt::<VcClaims>::from_str(&vc_b64).expect("VC as JWT");
        assert_snapshot!("vc_jwt", vc_jwt, {
            ".claims.iat" => "[iat]",
            ".claims.nbf" => "[nbf]",
            ".claims.vc.issuanceDate" => "[issuanceDate]",
            ".claims.vc.credentialSubject" => insta::sorted_redaction()
        });

        // token state should remain unchanged
        assert_let!(Ok(buf), StateManager::get(&provider, access_token).await);
        let state = State::try_from(buf).expect("token state is valid");
        assert_snapshot!("state", state, {
            ".expires_at" => "[expires_at]",
            ".token.c_nonce"=>"[c_nonce]",
            ".token.c_nonce_expires_at" => "[c_nonce_expires_at]"
        });

        // deferred state should not exist
        assert!(StateManager::get(&provider, transaction_id).await.is_err());
    }
}
