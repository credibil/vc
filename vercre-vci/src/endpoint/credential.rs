//! # Credential Handler
//!
//! The Credential Handler issues a Credential as approved by the End-User upon
//! presentation of a valid Access Token representing this approval.
//!
//! The Wallet sends one Credential Request per individual Credential to the
//! Credential Handler. The Wallet MAY use the same Access Token to send
//! multiple Credential Requests to request issuance of multiple Credentials of
//! different types bound to the same proof, or multiple Credentials of the same
//! type bound to different proofs.

use std::fmt::Debug;

use anyhow::anyhow;
use tracing::{instrument, trace};
use vercre_core::vci::{BatchCredentialRequest, CredentialRequest, CredentialResponse};
use vercre_core::{err, Callback, Client, Holder, Issuer, Result, Server, Signer, StateManager};

use super::Handler;

/// Credential request handler.
impl<P> Handler<P, CredentialRequest>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
{
    /// Call the request for the Request Object endpoint.
    #[instrument]
    pub async fn call(&self) -> Result<CredentialResponse> {
        trace!("Handler::call");
        self.handle_request(Context::new()).await
    }
}

#[derive(Debug)]
struct Context<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Callback + Clone,
{
    provider: Option<P>,
}

impl<P> Context<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Callback + Clone,
{
    #[instrument]
    pub fn new() -> Self {
        trace!("Context::new");
        Self { provider: None }
    }
}

impl<P> vercre_core::Context for Context<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone + Debug,
{
    type Provider = P;
    type Request = CredentialRequest;
    type Response = CredentialResponse;

    #[instrument]
    async fn init(&mut self, _: &Self::Request, provider: Self::Provider) -> Result<&Self> {
        trace!("Context::prepare");

        self.provider = Some(provider);
        Ok(self)
    }

    #[instrument]
    async fn process(&self, req: &Self::Request) -> Result<Self::Response> {
        trace!("Context::process");

        let Some(provider) = self.provider.clone() else {
            err!("Provider not set")
        };

        let batch_req = BatchCredentialRequest {
            credential_issuer: req.credential_issuer.clone(),
            access_token: req.access_token.clone(),
            credential_requests: vec![req.clone()],
        };
        let batch_res = Handler::new(provider, batch_req).call().await?;

        // set c_nonce and c_nonce_expires_at - batch endpoint sets them in the
        // top-level response, not each credential response
        let mut cred_res = batch_res.credential_responses[0].clone();
        cred_res.c_nonce = batch_res.c_nonce;
        cred_res.c_nonce_expires_in = batch_res.c_nonce_expires_in;

        Ok(cred_res)
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use assert_let_bind::assert_let;
    use base64ct::{Base64UrlUnpadded, Encoding};
    use chrono::Utc;
    use insta::assert_yaml_snapshot as assert_snapshot;
    use serde_json::json;
    use test_utils::vci_provider::{Provider, ISSUER, NORMAL_USER};
    use test_utils::wallet;
    use vercre_core::jwt::{self, Jwt};
    use vercre_core::vci::ProofClaims;
    use vercre_core::w3c::vc::VcClaims;

    use super::*;
    use crate::state::{Expire, State, TokenState};

    #[tokio::test]
    async fn credential_ok() {
        test_utils::init_tracer();

        let provider = &Provider::new();
        let access_token = "ABCDEF";
        let c_nonce = "1234ABCD";
        let credentials = vec!["EmployeeID_JWT".to_string()];

        // set up state
        let mut state = State::builder()
            .credential_issuer(ISSUER.to_string())
            .expires_at(Utc::now() + Expire::AuthCode.duration())
            .credentials(credentials)
            .holder_id(Some(NORMAL_USER.to_string()))
            .build();

        state.token = Some(TokenState {
            access_token: access_token.to_string(),
            token_type: "Bearer".to_string(),
            c_nonce: c_nonce.to_string(),
            c_nonce_expires_at: Utc::now() + Expire::Nonce.duration(),
            ..Default::default()
        });

        StateManager::put(&provider, access_token, state.to_vec(), state.expires_at)
            .await
            .expect("state saved");

        // create CredentialRequest to 'send' to the app
        let jwt_enc = Jwt {
            header: jwt::Header {
                typ: "vercre-vci-proof+jwt".to_string(),
                alg: wallet::alg(),
                kid: wallet::kid(),
            },
            claims: ProofClaims {
                iss: wallet::did(),
                aud: ISSUER.to_string(),
                iat: Utc::now().timestamp(),
                nonce: c_nonce.to_string(),
            },
        }
        .to_string();
        let sig = wallet::sign(jwt_enc.as_bytes());
        let sig_enc = Base64UrlUnpadded::encode_string(&sig);
        let signed_jwt = format!("{jwt_enc}.{sig_enc}");

        let body = json!({
            "credential_identifier": "EmployeeID_JWT",
            "proof":{
                "proof_type": "jwt",
                "jwt": signed_jwt
            }
        });

        let mut request =
            serde_json::from_value::<CredentialRequest>(body).expect("request should deserialize");
        request.credential_issuer = ISSUER.to_string();
        request.access_token = access_token.to_string();

        let response = Handler::new(provider, request).call().await.expect("response is valid");
        assert_snapshot!("response", response, {
            ".credential" => "[credential]",
            ".c_nonce" => "[c_nonce]",
            ".c_nonce_expires_in" => "[c_nonce_expires_in]"
        });

        // verify credential
        let vc_val = response.credential.expect("VC is present");
        let vc_b64 = serde_json::from_value::<String>(vc_val).expect("base64 encoded string");
        let vc_jwt = Jwt::<VcClaims>::from_str(&vc_b64).expect("VC as JWT");
        assert_snapshot!("vc_jwt", vc_jwt, {
            ".claims.iat" => "[iat]",
            ".claims.nbf" => "[nbf]",
            ".claims.vc.issuanceDate" => "[issuanceDate]",
            ".claims.vc.credentialSubject" => insta::sorted_redaction()
        });

        // token state should remain unchanged
        assert_let!(Ok(buf), StateManager::get(&provider, access_token).await);
        let state = State::try_from(buf).expect("state is valid");
        assert_snapshot!("state", state, {
            ".expires_at" => "[expires_at]",
            ".token.c_nonce"=>"[c_nonce]",
            ".token.c_nonce_expires_at" => "[c_nonce_expires_at]"
        });
    }
}
