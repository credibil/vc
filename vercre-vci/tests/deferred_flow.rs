use anyhow::Result;
use assert_let_bind::assert_let;
use base64ct::{Base64UrlUnpadded, Encoding};
use chrono::Utc;
use futures::future::TryFutureExt;
use insta::assert_yaml_snapshot as assert_snapshot;
use lazy_static::lazy_static;
use serde_json::json;
use test_utils::vci_provider::{Provider, ISSUER, PENDING_USER};
use test_utils::wallet;
use vercre_vci::credential::{CredentialRequest, CredentialResponse};
use vercre_vci::deferred::{DeferredCredentialRequest, DeferredCredentialResponse};
use vercre_vci::invoke::{InvokeRequest, InvokeResponse};
use vercre_vci::jwt::{self, Jwt};
use vercre_vci::token::{TokenRequest, TokenResponse};
use vercre_vci::{Endpoint, ProofClaims};

lazy_static! {
    static ref PROVIDER: Provider = Provider::new();
}

// Run through entire pre-authorized code flow.
#[tokio::test]
async fn deferred_flow() {
    test_utils::init_tracer();

    // go through the pre-auth flow to token endpoint
    let token_resp = get_offer().and_then(get_token).await.expect("Ok");
    let credential_resp = get_credential(token_resp.clone()).await.expect("Ok");

    assert_snapshot!("credential-response", &credential_resp, {
        ".transaction_id" => "[transaction_id]",
        ".c_nonce" => "[c_nonce]",
        ".c_nonce_expires_in" => "[c_nonce_expires_in]"
    });

    let deferred_resp = get_deferred(token_resp, credential_resp.clone()).await.expect("Ok");

    assert_snapshot!("deferred-response", deferred_resp, {
        ".credential" => "[credential]",
        ".c_nonce" => "[c_nonce]",
        ".c_nonce_expires_in" => "[c_nonce_expires_in]"
    });
}

// Simulate Issuer request to '/invoke' endpoint to get credential offer to use to
// make credential offer to Wallet.
async fn get_offer() -> Result<InvokeResponse> {
    // offer request
    let body = json!({
        "credential_configuration_ids": ["EmployeeID_JWT"],
        "holder_id": PENDING_USER,
        "pre-authorize": true,
        "tx_code_required": true,
        "callback_id": "1234"
    });

    let mut request = serde_json::from_value::<InvokeRequest>(body)?;
    request.credential_issuer = ISSUER.to_string();

    let endpoint = Endpoint::new(PROVIDER.to_owned());
    let response = endpoint.invoke(&request).await?;
    Ok(response)
}

// Simulate Wallet request to '/token' endpoint with pre-authorized code to get
// access token
async fn get_token(input: InvokeResponse) -> Result<TokenResponse> {
    assert_let!(Some(offer), &input.credential_offer);
    assert_let!(Some(grants), &offer.grants);
    assert_let!(Some(pre_authorized_code), &grants.pre_authorized_code);

    // create TokenRequest to 'send' to the app
    let body = json!({
        "client_id": wallet::did(),
        "grant_type": "urn:ietf:params:oauth:grant-type:pre-authorized_code",
        "pre-authorized_code": &pre_authorized_code.pre_authorized_code,
        "user_code": input.user_code.unwrap_or_default(),
    });

    let mut request = serde_json::from_value::<TokenRequest>(body)?;
    request.credential_issuer = ISSUER.to_string();

    let endpoint = Endpoint::new(PROVIDER.to_owned());
    let response = endpoint.token(&request).await?;
    Ok(response)
}

// Simulate Wallet request to '/credential' endpoint with access token to get credential.
async fn get_credential(input: TokenResponse) -> Result<CredentialResponse> {
    // create CredentialRequest to 'send' to the app
    let jwt_enc = Jwt {
        header: jwt::Header {
            typ: String::from("vercre-vci-proof+jwt"),
            alg: wallet::alg(),
            kid: wallet::kid(),
        },
        claims: ProofClaims {
            iss: wallet::did(),
            aud: ISSUER.to_string(),
            iat: Utc::now().timestamp(),
            nonce: input.c_nonce.unwrap_or_default(),
        },
    }
    .to_string();
    let sig = wallet::sign(jwt_enc.as_bytes());
    let sig_enc = Base64UrlUnpadded::encode_string(&sig);
    let signed_jwt = format!("{jwt_enc}.{sig_enc}");
    let body = json!({
        "format": "jwt_vc_json",
        "credential_definition": {
            "type": [
                "VerifiableCredential",
                "EmployeeIDCredential"
            ]
        },
        "proof":{
            "proof_type": "jwt",
            "jwt": signed_jwt
        }
    });

    let mut request = serde_json::from_value::<CredentialRequest>(body)?;
    request.credential_issuer = ISSUER.to_string();
    request.access_token = input.access_token;

    let endpoint = Endpoint::new(PROVIDER.to_owned());
    let response = endpoint.credential(&request).await?;
    Ok(response)
}

async fn get_deferred(
    tkn: TokenResponse, cred_resp: CredentialResponse,
) -> Result<DeferredCredentialResponse> {
    let request = DeferredCredentialRequest {
        credential_issuer: ISSUER.to_string(),
        access_token: tkn.access_token,
        transaction_id: cred_resp.transaction_id.expect("should have transaction_id"),
    };

    let endpoint = Endpoint::new(PROVIDER.to_owned());
    let response = endpoint.deferred(&request).await?;
    Ok(response)
}
