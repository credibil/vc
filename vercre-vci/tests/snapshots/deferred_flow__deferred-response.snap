---
source: vercre-vci/tests/deferred_flow.rs
assertion_line: 39
expression: deferred_resp
---
format: jwt_vc_json
credential: "[credential]"
c_nonce: "[c_nonce]"
c_nonce_expires_in: "[c_nonce_expires_in]"

